### PR TITLE
16 implement category entity and category use cases

### PR DESCRIPTION
**Pull request by EmilyDo**
Description: This pull request implements the Category and CategoryFactory entities, 3 use cases, (CreateCategory, DeleteCategory, and EditCategory), and UI files for To Do List model, CategoryList, and CreateCategory, and entity unit tests.

CreateCategory: creating a new Category object and adding it to Category Factory.
DeleteCategory: deleting an existing Category object and removing it from Category Factory.
EditCategory: modifying attributes of a Category object, e.g. name of the Category, colour of the Category, or visibility state of the Category.

Designs:
- I created separate files for the UI files: TitleBar (responsibility = creating a title panel with a title and buttons), CategoryList (responsibility = creating a new panel for a new category), and ToDoViewModel (the To Do list view) to ensure Single Responsibility Principle of SOLID.
- Used interfaces for Boundary files to ensure the Open/Close principle of SOLID. 

Design Pattern: 
- An abstract Factory (implemented on CategoryFactory) 
- Façade (for use cases)  

- [ ] Unit tests: About 10%. Only unit tests for the Category entity are finished.
- [ ] Exception Handling: Incomplete
- [x] JavaDocs Documentation: Documentations are added to most of the functions, except for some methods
- [x] SOLID Principles: Followed Single Responsibility Principle, all files except UI package files follow the Open/Close principle, and Interface Segregation. Need to improve Dependency Inversion, use case interactor should be passed the interface as an argument, not the presenter (or else it wouldn’t invert depency). 
- [x] Clean Architecture: The structure partially follows Clean Architecture.

### CODE REVIEW

File: src/test/java/testentities/TestCategory.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+public class TestCategory {
+
+    Category category = new Category("name", "colour");
+    @Test(timeout = 500)
+    public void testConstructor() {
+        String name = "Chores";
+        String colour = "RED";
+        Category example = new Category(name, colour);
+        // assertion
```
Comment (JoAndNight) [inline src/test/java/testentities/TestCategory.java:15]: This kind of comment is not necessary.

File: src/test/java/test_use_cases/TestCategory.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+package test_use_cases;
+
+import entities.Category;
+import entities.CategoryFactory;
+import use_cases.categories.create_category.*;
+
+import static org.junit.Assert.*;
+
+public class TestCategory {
```
Comment (JoAndNight) [inline src/test/java/test_use_cases/TestCategory.java:11]: The name of the class is the same as the name of another class in testentities package. You need to rename it properly.

File: src/test/java/test_use_cases/TestCategory.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+     CategoryFactory factory = new CategoryFactory();
+     String name = "Chores";
+     String colour = "RED";
+     CreateCategoryOutputBound outputBound = new CreateCategoryOutputBound() {
+        @Override
+        public CreateCategoryOutputData prepareSuccessView(CreateCategoryOutputData outputData) {
+            assertEquals(name, outputData.getCategory().getName());
```
Comment (JoAndNight) [inline src/test/java/test_use_cases/TestCategory.java:18]: Adding an assertion message is recommended.

File: src/test/java/test_use_cases/TestCategory.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            assertTrue(factory.contains(outputData.getCategory().getName(), false));
+            return null;
+        }
+        @Override
+        public CreateCategoryOutputData prepareFailView(CreateCategoryOutputData outputData) {
+            return null;
+        }
+    } ;
+
+    CreateCategoryInputBound inputBound = new CreateCategoryInputBound() {
```
Comment (JoAndNight) [inline src/test/java/test_use_cases/TestCategory.java:28]: I don't really understand the lines below. Can you explain them?

File: src/main/java/use_cases/categories/create_category/CreateCategory.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
+package use_cases.categories.create_category;
+import entities.Category;
+import entities.CategoryFactory;
+
+public class CreateCategory implements CreateCategoryInputBound {
+    private final CreateCategoryPresenter presenter;
```
[the remark this answers is not in the document]
Comment (EmilyDo) [reply src/main/java/use_cases/categories/create_category/CreateCategory.java:6]: OutputBoundary is an interface, how do we add that as a parameter without having to override all the abstract methods everytime we call on a use case constructor?

[the remark this answers is not in the document]
Comment (AngelEyesMagicMaker) [reply src/main/java/use_cases/categories/create_category/CreateCategory.java:6]: Assign a specific implementation to that field using the class constructor.

Look for constructor usages to check where it is called from. Hence you will know from which place that implementation comes.

Hope this help.

**Review by m-banthwan — DISMISSED** [review]

**Review by JoAndNight — APPROVED** [review]

**Review by AngelEyesMagicMaker — APPROVED** [review]

**Review by m-banthwan — APPROVED** [review]

**Pull request merged**